### PR TITLE
Major parsing cleanup.

### PR DESCRIPTION
**Pull request by kormat**
Add lib.util.Raw, to wrap raw bytes, make offset accounting automatic,
and make parsing code much easier to follow.

All parsing errors should now raise an appropriate exception.

This follows on from #272.
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/netsec-ethz/scion/pull/276%23issuecomment-125601429%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/276%23issuecomment-125603756%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/276%23issuecomment-125919917%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/276%23discussion_r35750329%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/276%23discussion_r35750931%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/276%23discussion_r35750956%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/276%23discussion_r35750993%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/276%23discussion_r35750998%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/276%23discussion_r35751118%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/276%23issuecomment-125934068%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/276%23issuecomment-125935311%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/276%23issuecomment-125601429%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22The%20next%20step%20would%20be%20to%20immediately%20wrap%20all%20%60bytes%60%20values%20in%20%60Raw%60%2C%20and%20only%20pass%20around%20%60Raw%60%20objects.%22%2C%20%22created_at%22%3A%20%222015-07-28T13%3A10%3A16Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22This%20breaks%20end2end%2C%20currently%20debugging.%22%2C%20%22created_at%22%3A%20%222015-07-28T13%3A18%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22Bugs%20found%20and%20fixed.%20This%20PR%20is%20ready%20for%20review%20now.%20All%20unit%20tests%20pass%2C%20end2end%20works.%22%2C%20%22created_at%22%3A%20%222015-07-29T11%3A00%3A51Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22lgtm%22%2C%20%22created_at%22%3A%20%222015-07-29T12%3A18%3A10Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22If%20we%20decide%20to%20go%20with%20this%20approach%2C%20LGTM%22%2C%20%22created_at%22%3A%20%222015-07-29T12%3A25%3A39Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3840858%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/shitz%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%2073672af98bb7540ef8cbcab2dd5e271a86322d58%20lib/util.py%2081%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/276%23discussion_r35750956%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22whitespaces%20around%20%27%2B%27%22%2C%20%22created_at%22%3A%20%222015-07-29T11%3A57%3A08Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3840858%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/shitz%22%7D%7D%2C%20%7B%22body%22%3A%20%22Fixed.%22%2C%20%22created_at%22%3A%20%222015-07-29T12%3A00%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/util.py%3AL331-430%22%7D%2C%20%22Pull%2073672af98bb7540ef8cbcab2dd5e271a86322d58%20lib/util.py%20110%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/276%23discussion_r35750931%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20know%20you%20don%27t%20like%20properties%2C%20but%20for%20consistency%20reason%20this%20should%20be%20decorated%20with%20%60%40property%60.%22%2C%20%22created_at%22%3A%20%222015-07-29T11%3A56%3A42Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3840858%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/shitz%22%7D%7D%2C%20%7B%22body%22%3A%20%22No%20%3AP%22%2C%20%22created_at%22%3A%20%222015-07-29T11%3A58%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/util.py%3AL331-430%22%7D%2C%20%22Pull%2073672af98bb7540ef8cbcab2dd5e271a86322d58%20lib/packet/path_mgmt.py%20253%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/276%23discussion_r35750329%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22why%20not%20%60pop%28%29%60%3F%22%2C%20%22created_at%22%3A%20%222015-07-29T11%3A46%3A36Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22It%27s%20the%20only%20type%20with%20a%20known%2C%20fixed%20length.%22%2C%20%22created_at%22%3A%20%222015-07-29T11%3A57%3A56Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/packet/path_mgmt.py%3AL471-492%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/276#issuecomment-125601429'>General Comment</a></b>
- <a href='https://github.com/kormat'></a> The next step would be to immediately wrap all `bytes` values in `Raw`, and only pass around `Raw` objects.
- <a href='https://github.com/kormat'></a> This breaks end2end, currently debugging.
- <a href='https://github.com/kormat'></a> Bugs found and fixed. This PR is ready for review now. All unit tests pass, end2end works.
- <a href='https://github.com/pszalach'></a> lgtm
- <a href='https://github.com/shitz'></a> If we decide to go with this approach, LGTM
- [ ] <a href='#crh-comment-Pull 73672af98bb7540ef8cbcab2dd5e271a86322d58 lib/packet/path_mgmt.py 253'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/276#discussion_r35750329'>File: lib/packet/path_mgmt.py:L471-492</a></b>
- <a href='https://github.com/pszalach'></a> why not `pop()`?
- <a href='https://github.com/kormat'></a> It's the only type with a known, fixed length.
- [ ] <a href='#crh-comment-Pull 73672af98bb7540ef8cbcab2dd5e271a86322d58 lib/util.py 110'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/276#discussion_r35750931'>File: lib/util.py:L331-430</a></b>
- <a href='https://github.com/shitz'></a> I know you don't like properties, but for consistency reason this should be decorated with `@property`.
- <a href='https://github.com/kormat'></a> No :P
- [ ] <a href='#crh-comment-Pull 73672af98bb7540ef8cbcab2dd5e271a86322d58 lib/util.py 81'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/276#discussion_r35750956'>File: lib/util.py:L331-430</a></b>
- <a href='https://github.com/shitz'></a> whitespaces around '+'
- <a href='https://github.com/kormat'></a> Fixed.

<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/276?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/276?approve=1'></a>&nbsp;<a href='https://github.com/netsec-ethz/scion/pull/276'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: lib/packet/path_mgmt.py [hunk 1]
```diff
@@ -473,19 +471,21 @@ def __init__(self, raw=None):
...
     def parse(self, raw):
         SCIONPacket.parse(self, raw)
+        data = Raw(self.payload, "PathMgmtPacket", self.MIN_LEN, min_=True)
         # Get the type of the first byte of the payload and instantiate the
         # corresponding payload class.
-        self.type = struct.unpack("!B", self.payload[0:1])[0]
+        self.type = data.pop(1)
         if self.type == PathMgmtType.REQUEST:
-            self.payload = PathSegmentInfo(self.payload[1:])
+            self.payload = PathSegmentInfo(data.pop(PathSegmentInfo.LEN))
```
Comment (pszal) [inline lib/packet/path_mgmt.py:479]: why not `pop()`?

File: lib/util.py [hunk 2]
```diff
@@ -325,3 +331,99 @@ def set_time_method(cls, method=None):
...
+            self._offset = dlen
+        elif n == 1:
+            self._offset += 1
+        else:
+            self._offset += n
+        if self._offset > dlen:
+            self._offset = dlen
+        return ret
+
+    def offset(self):
```
Comment (shitz) [inline lib/util.py:425]: I know you don't like properties, but for consistency reason this should be decorated with `@property`.

File: lib/util.py [hunk 3]
```diff
@@ -325,3 +331,99 @@ def set_time_method(cls, method=None):
...
+
+        If `n` is not specified, return all remaining elements of data.
+        If `n` is 1, return the next element of data (as an int).
+        If `n` is > 1, return the next `n` elements of data (as bytes).
+
+        :param n: How many elements to return (see above)
+        :param bool bounds: Perform bounds checking on access if True
+        """
+        dlen = len(self._data)
+        if n and bounds and self._offset+n > dlen:
```
Comment (shitz) [inline lib/util.py]: whitespaces around '+'

File: lib/packet/path_mgmt.py (hunk 1 above)
Comment (kormat) [inline lib/packet/path_mgmt.py:479]: It's the only type with a known, fixed length.

File: lib/util.py (hunk 2 above)
Comment (kormat) [inline lib/util.py:425]: No :P

File: lib/util.py (hunk 3 above)
Comment (kormat) [inline lib/util.py]: Fixed.

**Pull request merged**